src/App.js

```diff
@@ -26,9 +26,35 @@ const generateSquares = () => {
 };
 
 const App = () => {
-  // This starts state off as a 2D array of JS objects with
+  // This state starts off as a 2D array of JS objects with
   // empty value and unique ids.
-  const [squares, setSquares] = useState(generateSquares());
+  const [squares, updateSquares] = useState(generateSquares());
+  const [currentPlayer, updateCurrentPlayer] = useState(PLAYER_1);
+  const [winner, updateWinner] = useState(null);
+
+  // when the square state changes, check to see if there is a winner
+  // (calling checkForWinner() in updateSquare for seem reason always was behind in state)
+  useEffect(() => {
```

Comment:
Nice solution!

Reply:
I believe what's going on here is related to the fact that setting state is asynchronous